include/picongpu/particles/Particles.tpp

```diff
@@ -48,12 +48,144 @@
 #include <memory>
-
+#include <utility>
 
 namespace picongpu
 {
     using namespace pmacc;
 
+    namespace detail
+    {
+        /* Helper to check if a member exists
+         *
+         * Derived from C++17 std::void_t.
+         * This implementation will be removed with Void provided by alpaka 0.6.0 release (not included in the 0.6.0rc3
+         * we currently using).
+         */
+        template<class...>
+        using Void = void;
+
+        /** Calculate the scaling factor for each direction.
+         *
+         * The scaling factor is derived from the reference size of the local domain and a scaling factor provided by
+         * the user.
+         *
+         * @tparam T_ExchangeMemCfg exchange configuration for a species
+         * @tparam T_Sfinae Type for conditionally specialization (no input parameter)
+         * @{
+         */
+        template<typename T_ExchangeMemCfg, typename T_Sfinae = void>
+        struct DirScalingFactor
+        {
+            //! @return factor to scale the amount of memory for each direction
+            static floatD_64 get()
+            {
+                return floatD_64::create(1.0);
+            }
+        };
+
+        /** Specialization for species with exchange memory information which provides
+         * DIR_SCALING_FACTOR and REF_LOCAL_DOM_SIZE
+         */
+        template<typename T_ExchangeMemCfg>
+        struct DirScalingFactor<
+            T_ExchangeMemCfg,
+            Void<
+                decltype(std::declval<T_ExchangeMemCfg>().DIR_SCALING_FACTOR),
+                typename T_ExchangeMemCfg::REF_LOCAL_DOM_SIZE>>
+        {
+            static floatD_64 get()
+            {
+                auto baseLocalCells = T_ExchangeMemCfg::REF_LOCAL_DOM_SIZE::toRT();
+                auto userScalingFactor = T_ExchangeMemCfg{}.DIR_SCALING_FACTOR;
+
+                auto localDomSize = Environment<simDim>::get().SubGrid().getLocalDomain().size;
+                // set too local domain size in case there is no base volume defined
+                for(uint32_t d = 0; d < simDim; ++d)
+                {
+                    if(baseLocalCells[d] <= 0)
+                        baseLocalCells[d] = localDomSize[d];
+                }
+
+                auto scale = floatD_64::create(1.0);
+                for(uint32_t d = 0; d < simDim; ++d)
+                {
+                    auto dir1 = (d + 1) % simDim;
+                    auto dir2 = (d + 2) % simDim;
+                    // precision: numbers are small, therefore the usage of double is fine
+                    auto scaleDirection = std::ceil(
+                        float_64(localDomSize[dir1]) / float_64(baseLocalCells[dir1]) * float_64(localDomSize[dir2])
+                        / float_64(baseLocalCells[dir2]));
+                    float_64 scalingFactor = scaleDirection * userScalingFactor[d];
+                    // do not scale down
+                    scale[d] = scalingFactor > 1.0 ? scalingFactor : 1.0;
```

Comment:
This can be expressed as just `max(scalingFactor, 1.0)`?

Reply:
I was too lazy to include `<algorithm>` ^^

Reply:
add `std::max`